tools/stats/upload_test_stats.py

```diff
@@ -0,0 +1,181 @@
+    return ret
+
+
+def get_artifact_urls(workflow_run_id: int) -> Dict[Path, str]:
+    """Get all workflow artifacts with 'test-report' in the name."""
+    response = requests.get(
+        f"{PYTORCH_REPO}/actions/runs/{workflow_run_id}/artifacts?per_page=100",
+    )
+    artifacts = response.json()["artifacts"]
+    while "next" in response.links.keys():
+        response = requests.get(response.links["next"]["url"], headers=REQUEST_HEADERS)
+        artifacts.extend(response.json()["artifacts"])
+
+    artifact_urls = {}
+    for artifact in artifacts:
+        if "test-report" in artifact["name"]:
+            artifact_urls[Path(artifact["name"])] = artifact["archive_download_url"]
+    return artifact_urls
+
+
+def unzip(p: Path) -> None:
+    """Unzip the provided zipfile to a similarly-named directory.
+
+    Returns None if `p` is not a zipfile.
+
+    Looks like: /tmp/test-reports.zip -> /tmp/unzipped-test-reports/
+    """
+    assert p.is_file()
+    unzipped_dir = p.with_name("unzipped-" + p.stem)
+
+    with zipfile.ZipFile(p, "r") as zip:
+        zip.extractall(unzipped_dir)
+
+
+def download_and_extract_artifact(artifact_name: Path, artifact_url: str) -> None:
+    response = requests.get(artifact_url, headers=REQUEST_HEADERS)
+    print(f"Downloading and extracting {artifact_name}")
+    with open(artifact_name, "wb") as f:
+        f.write(response.content)
+    unzip(artifact_name)
+
+
+def download_and_extract_s3_reports(workflow_run_id: int) -> None:
+    bucket = S3_RESOURCE.Bucket("gha-artifacts")
+    objs = bucket.objects.filter(
+        Prefix=f"pytorch/pytorch/{workflow_run_id}/artifact/test-reports"
+    )
+
+    for obj in objs:
+        p = Path(Path(obj.key).name)
+        print(f"Downloading and extracting {p}")
+        with open(p, "wb") as f:
+            f.write(obj.get()["Body"].read())
+        unzip(p)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Upload test stats to Rockset")
+    parser.add_argument(
+        "--workflow-run-id",
+        required=True,
+        help="id of the workflow to get artifacts from",
+    )
+    args = parser.parse_args()
+
+    print("mkdir: ", TEMP_DIR)
+    TEMP_DIR.mkdir()
+    print("cd to ", TEMP_DIR)
+    os.chdir(TEMP_DIR)
+
+    # Download and extract all the reports (both GHA and S3)
+    download_and_extract_s3_reports(args.workflow_run_id)
+    artifact_urls = get_artifact_urls(args.workflow_run_id)
+    for name, url in artifact_urls.items():
+        download_and_extract_artifact(Path(name), url)
+
+    # Parse the reports and transform them to JSON
+    test_cases = []
+    for xml_report in Path(".").glob("**/*.xml"):
+        test_cases.extend(parse_xml_report(xml_report, int(args.workflow_run_id)))
+
+    # Write the JSON to rockset
+    print(f"Writing {len(test_cases)} test cases to Rockset")
+    client = rockset.Client(
+        api_server="api.rs2.usw2.rockset.com", api_key=os.environ["ROCKSET_API_KEY"]
```

Comment:
should we move this constant to "api.rs2.usw2.rockset.com" to beginning of the file with other constants ?